src/policy/policy-inl.h

```diff
@@ -0,0 +1,196 @@
+    SetRecursively(Permissions::k##name, value);
+void Policy::SetRecursively(Permissions permission, bool value) {
+  if (permission != Permissions::kPermissionsRoot)
+    permissions_[static_cast<size_t>(permission)] = value;
+  PERMISSIONS(V)
+}
+#undef V
+
+template <typename...P>
+bool Policy::granted(P...permissions) const {
+  std::array<Permissions, sizeof...(permissions)> perms {
+    std::forward<P>(permissions)...
+  };
+  for (Permissions permission : perms) {
+    if (UNLIKELY(!test(permission)))
+      return false;
+  }
+  return true;
+}
+
+template <typename...P>
+void Policy::grant(P... permissions) {
+  if (locked_) return;
+  std::array<Permissions, sizeof...(permissions)> perms {
+    std::forward<P>(permissions)...
+  };
+  for (Permissions permission : perms)
+    SetRecursively(permission, false);
+}
+
+template <typename...P>
+void Policy::deny(P... permissions) {
+  std::array<Permissions, sizeof...(permissions)> perms {
+    std::forward<P>(permissions)...
+  };
+  for (Permissions permission : perms)
+    SetRecursively(permission, true);
+}
+
+#define V(Name, label, _)                                                      \
+  if (strcmp(name.c_str(), label) == 0) return Permissions::k##Name;
+Permissions Policy::PermissionFromName(const std::string& name) {
+  if (strcmp(name.c_str(), "*") == 0) return Permissions::kPermissionsRoot;
+  PERMISSIONS(V)
+  return Permissions::kPermissionsCount;
+}
+#undef V
+
+template <typename...P>
+bool Policy::grantedByName(P...permissions) const {
+  std::array<std::string, sizeof...(permissions)> perms {
+    std::forward<P>(permissions)...
+  };
+  for (std::string& name : perms) {
+    Permissions permission = PermissionFromName(name);
+    if (permission == Permissions::kPermissionsCount || !granted(permission))
+      return false;
+  }
+  return true;
+}
+
+template <typename...P>
+void Policy::grantByName(P... permissions) {
+  if (locked_) return;
+  std::array<std::string, sizeof...(permissions)> perms {
+    std::forward<P>(permissions)...
+  };
+  for (std::string& name : perms) {
+    Permissions permission = PermissionFromName(name);
+    if (permission != Permissions::kPermissionsCount)
+      grant(permission);
+  }
+}
+
+template <typename...P>
+void Policy::denyByName(P... permissions) {
+  if (locked_) return;
+  std::array<std::string, sizeof...(permissions)> perms {
+    std::forward<P>(permissions)...
+  };
+  for (std::string& name : perms) {
+    Permissions permission = PermissionFromName(name);
+    if (permission != Permissions::kPermissionsCount)
+      deny(permission);
+  }
+}
+
+std::vector<Permissions> Policy::Parse(
+    const std::string& permissions,
+    std::vector<std::string>* errors) {
+  std::vector<Permissions> perms;
+  std::string::size_type pos = 0;
+  std::string::size_type index = 0;
+  Permissions perm;
+  std::string name;
+
+  while ((index = permissions.find_first_of(",", pos)) != std::string::npos) {
```

Comment:
We have `SplitString` in `util.h`, maybe that’s a bit nicer to work with?

(I also feel like this could go into the `.cc` file, it’s not really very inline-able)

Reply:
I was pretty certain we had something but for the life of me I couldn't remember where it was or what it was called 😆 ... will definitely switch to `SplitString` !